src/Worker/Delivery.php

```diff
@@ -345,175 +167,245 @@ public static function execute($cmd, $item_id, $contact_id) {
-					break;
-				} elseif ($top_level && !$walltowall) {
-					// currently no workable solution for sending walltowall
-					logger('diaspora status: '.$loc);
-					Diaspora::sendStatus($target_item,$owner,$contact,$public_message);
-					break;
-				}
+		if (($deliver_status >= 200) && ($deliver_status <= 299)) {
+			// We successfully delivered a message, the contact is alive
+			Contact::unmarkForArchival($contact);
+		} else {
+			// The message could not be delivered. We mark the contact as "dead"
+			Contact::markForArchival($contact);
+		}
+	}
 
-				logger('delivery: diaspora unknown mode: '.$contact['name']);
+	private static function deliverDiaspora($cmd, $contact, $owner, $items, $target_item, $public_message, $top_level, $followup)
+	{
+		// We don't treat Forum posts as "wall-to-wall" to be able to post them via Diaspora
+		$walltowall = $top_level && ($owner['id'] != $items[0]['contact-id']) & ($owner['account-type'] != ACCOUNT_TYPE_COMMUNITY);
 
-				break;
+		if ($public_message) {
+			$loc = 'public batch ' . $contact['batch'];
+		} else {
+			$loc = $contact['addr'];
+		}
 
-			default:
-				break;
+		logger('Deliver ' . $target_item["guid"] . ' via Diaspora to ' . $loc);
+
+		if (Config::get('system', 'dfrn_only') || !Config::get('system', 'diaspora_enabled')) {
+			return;
+		}
+		if ($cmd == DELIVER_MAIL) {
+			Diaspora::sendMail($target_item, $owner, $contact);
+			return;
 		}
 
-		return;
+		if ($cmd == DELIVER_SUGGESTION) {
+			return;
+		}
+		if (!$contact['pubkey'] && !$public_message) {
+			return;
+		}
+		if (($target_item['deleted']) && (($target_item['uri'] === $target_item['parent-uri']) || $followup)) {
+			// top-level retraction
+			logger('diaspora retract: ' . $loc);
+			Diaspora::sendRetraction($target_item, $owner, $contact, $public_message);
+			return;
+		} elseif ($cmd == DELIVER_RELOCATION) {
+			Diaspora::sendAccountMigration($owner, $contact, $owner['uid']);
+			return;
+		} elseif ($followup) {
+			// send comments and likes to owner to relay
+			logger('diaspora followup: ' . $loc);
+			Diaspora::sendFollowup($target_item, $owner, $contact, $public_message);
+			return;
+		} elseif ($target_item['uri'] !== $target_item['parent-uri']) {
+			// we are the relay - send comments, likes and relayable_retractions to our conversants
+			logger('diaspora relay: ' . $loc);
+			Diaspora::sendRelay($target_item, $owner, $contact, $public_message);
+			return;
+		} elseif ($top_level && !$walltowall) {
+			// currently no workable solution for sending walltowall
+			logger('diaspora status: ' . $loc);
+			Diaspora::sendStatus($target_item, $owner, $contact, $public_message);
+			return;
+		}
+
+		logger('Unknown mode ' . $cmd . ' for ' . $loc);
 	}
 
-	private static function getItemContact($item, $contacts)
+	private static function deliverMail($cmd, $contact, $owner, $target_item)
 	{
-		if (!count($contacts) || !is_array($item)) {
-			return false;
+		global $a;
```

Comment:
Please make `Worker\Delivery` extends `BaseObject` and call `$a = self::getApp()` instead. The goal is to remove the global `$a` variable.